apps/emqx/src/emqx_tls_lib.erl

```diff
@@ -556,33 +556,96 @@ resolve_cert_path_for_read_strict(Path) ->
 resolve_cert_path_for_read(Path) ->
     emqx_schema:naive_env_interpolation(Path).
 
-filter([], _) ->
-    [];
-filter([{_, undefined} | T], Versions) ->
-    filter(T, Versions);
-filter([{_, ""} | T], Versions) ->
-    filter(T, Versions);
-filter([{K, V} | T], Versions) ->
+ensure_valid_options(Options, Versions0) ->
+    Versions = validate_version_gap(Versions0),
```

Comment:
`Versions` are updated here and used in the next function call to validate other opts, but returned Options will still hold original `Versions`. 
I suggest moving version gap validation to emqx_schema, so we will disallow listener create/update with the version gap and return an error message to a user if it happens.

Reply:
moved to emqx_schema